includes/profile.inc

```diff
@@ -449,3 +459,368 @@ function cm_tools_update_missing_default_field_values_insert_with_languages(arra
+  }
+
+  switch ($s['state']) {
+
+    case 'GATHER ROWS':
+      if (!isset($s['table'])) {
+        $s['state'] = 'END';
+        break;
+      }
+      list($table_name, $primary_columns) = $s['table'];
+      $query = db_select($table_name, 'f')
+        ->fields('f', $primary_columns);
+      if (is_callable($query_alter_callback)) {
+        call_user_func($query_alter_callback, $query, $table_name);
+      }
+      $s['rows'] = $query
+        ->execute()
+        ->fetchAll(PDO::FETCH_ASSOC);
+      $s['state'] = 'PROCESS ROWS';
+      $status[] = strtr('Found !row_count rows in !table_name.', array(
+        '!row_count' => count($s['rows']),
+        '!table_name' => $table_name,
+      ));
+      // Deliberately fall-through, because why not?
+      //break;
+
+    case 'PROCESS ROWS':
+      list($table_name, $primary_columns) = $s['table'];
+      $rows = array_splice($s['rows'], 0, $rows_per_pass);
+      if (empty($rows)) {
+        $status[] = strtr('Updated !row_count rows in !table_name.', array(
+          '!row_count' => $s['rows_updated'],
+          '!table_name' => $table_name,
+        ));
+        $s['rows_updated'] = 0;
+        $s['state'] = 'GATHER ROWS';
+        $s['table'] = $move_to_next_table();
+        break;
+      }
+      // Fetch the values for all the rows in our chunk in the same query.
+      $qry = db_select($table_name, 'f')
+        ->fields('f');
+      $or = db_or();
+      foreach ($rows as $row) {
+        $and = db_and();
+        foreach ($row as $primary_column => $value) {
+          $and->condition($primary_column, $value);
+        }
+        $or->condition($and);
+      }
+      $rows = $qry
+        ->condition($or)
+        ->execute()
+        ->fetchAll(PDO::FETCH_ASSOC);
+      foreach ($rows as $original_row) {
+        $altered_row = $original_row;
+        $return = call_user_func_array($work_callback, array(
+          &$altered_row,
+          $table_name,
+        ));
+        if ($return === FALSE) {
+          continue;
+        }
+        // If a non-primary column changed, update the DB record.
+        $changed_columns = array();
+        foreach ($altered_row as $column => $value) {
+          if (!in_array($column, $primary_columns)) {
+            if ($return === TRUE || ($altered_row[$column] !== $original_row[$column])) {
+              $changed_columns[$column] = $altered_row[$column];
+            }
+          }
+        }
+        if (!empty($changed_columns)) {
+          $qry = db_update($table_name)
+            ->fields($changed_columns);
+          foreach ($original_row as $column => $value) {
+            if (in_array($column, $primary_columns)) {
+              $qry->condition($column, $value);
+            }
+          }
+          $s['rows_updated'] += $qry->execute();
+        }
+      }
+      break;
+  }
+
+  if ($s['state'] === 'END') {
+    $sandbox['#finished'] = 1;
+  }
+  else {
+    $sandbox['#finished'] = 0.5;
```

Comment:
It feels a slight shame that this can't be a more accurate number to indicate progress, but I can also imagine that would be hard to make useful/accurate anyway, so fair enough.

Reply:
I did have that in an earlier version of the code, but then I made the callback return the next table only, so it doesn't really have the information about the total number of tables anyway. Also, do we ever really see a report of the current progress of an update hook?